### ISSUE TITLE
[networkd] dhcp4: anonymize=true needs better documentation

### ISSUE THREAD

**Issue by MelkorLord** [systemd/systemd#11551]
**systemd version the issue has been seen with**
> 237-3ubuntu10.11

**Used distribution**
 > Ubuntu 18.04 Bionic Beaver

**Expected behaviour you didn't see**
 > Sending minimal DHCP options to the DHCP server

**Unexpected behaviour you saw**
 > Sending more options to the DHCP server with `[DHCP] Anonymize=true` than when not used (which is equivalent to `Anonymize=false` according to the docs

**Steps to reproduce the problem**
Using `dhcpdump` on the DHCP server, here are the results with the following systemd-networkd configuration:

/etc/systemd/network/50-eth0.network
```
[Match]
name=eth0

[Network]
DHCP=ipv4

[DHCP]
Anonymize=true
```
Here is the DHCPDISCOVER sent : 
```
  TIME: 2019-01-25 02:41:42.523
    IP: 0.0.0.0 (80:ee:73:40:6c:ad) > 255.255.255.255 (ff:ff:ff:ff:ff:ff)
    OP: 1 (BOOTPREQUEST)
 HTYPE: 1 (Ethernet)
  HLEN: 6
  HOPS: 0
   XID: 84be50e4
  SECS: 1
 FLAGS: 0
CIADDR: 0.0.0.0
YIADDR: 0.0.0.0
SIADDR: 0.0.0.0
GIADDR: 0.0.0.0
CHADDR: 80:ee:73:40:6c:ad:00:00:00:00:00:00:00:00:00:00
 SNAME: .
 FNAME: .
OPTION:  53 (  1) DHCP message type         1 (DHCPDISCOVER)
OPTION:  61 (  7) Client-identifier         01:80:ee:73:40:6c:ad
OPTION:  55 ( 14) Parameter Request List      1 (Subnet mask)
                                              3 (Routers)
                                              6 (DNS server)
                                             15 (Domainname)
                                             31 (Perform router discovery)
                                             33 (Static route)
                                             43 (Vendor specific info)
                                             44 (NetBIOS name server)
                                             46 (NetBIOS node type)
                                             47 (NetBIOS scope)
                                            121 (Classless Static Route)
                                            249 (MSFT - Classless route)
                                            252 (MSFT - WinSock Proxy Auto Detect)
                                             42 (NTP servers)
```
That's a **lot** of information to request IMHO!

Now, I commented the whole `[DHCP]` section and restarted `systemd-networkd.service` and captured again : 

```
  TIME: 2019-01-25 02:44:21.571
    IP: 0.0.0.0 (80:ee:73:40:6c:ad) > 255.255.255.255 (ff:ff:ff:ff:ff:ff)
    OP: 1 (BOOTPREQUEST)
 HTYPE: 1 (Ethernet)
  HLEN: 6
  HOPS: 0
   XID: 0ba12b5e
  SECS: 1
 FLAGS: 0
CIADDR: 0.0.0.0
YIADDR: 0.0.0.0
SIADDR: 0.0.0.0
GIADDR: 0.0.0.0
CHADDR: 80:ee:73:40:6c:ad:00:00:00:00:00:00:00:00:00:00
 SNAME: .
 FNAME: .
OPTION:  53 (  1) DHCP message type         1 (DHCPDISCOVER)
OPTION:  61 ( 19) Client-identifier         ff:cb:39:0a:c7:00:02:00:00:ab:11:59:61:68:2c:51:07:a6:99
OPTION:  55 (  8) Parameter Request List      1 (Subnet mask)
                                              3 (Routers)
                                             12 (Host name)
                                             15 (Domainname)
                                              6 (DNS server)
                                             33 (Static route)
                                            121 (Classless Static Route)
                                             42 (NTP servers)
............................................
OPTION:  57 (  2) Maximum DHCP message size 576
OPTION:  12 (  6) Host name                 testhost
```
The difference is quite big! `Anonymize=false` sends **less** request parameters to the DHCP server. Granted, the request is anonymized but why so many parameters are requested?

Is this expected behavior?

Moreover, I see here security issues. This test was against a local, well behaved DHCP server. Those people unfortunate enough to deal with not-so-nice ISPs would see their host receiving (and perhaps auto-setting I haven't tested that) lots of unrequested/unexpected settings like **NETBIOS** and **Proxy Auto Detect** information! In my opinion, this should raise eyebrows!

I've modified the `.network` file to mimic the behavior I wanted for a minimal DHCP client request and added the following : 
```
[DHCP]
SendHostname=false
UseHostname=false
UseNTP=false
ClientIdentifier=mac
```
and this resulted in : 
```
  TIME: 2019-01-25 03:20:44.755
    IP: 0.0.0.0 (80:ee:73:40:6c:ad) > 255.255.255.255 (ff:ff:ff:ff:ff:ff)
    OP: 1 (BOOTPREQUEST)
 HTYPE: 1 (Ethernet)
  HLEN: 6
  HOPS: 0
   XID: 78ed8781
  SECS: 1
 FLAGS: 0
CIADDR: 0.0.0.0
YIADDR: 0.0.0.0
SIADDR: 0.0.0.0
GIADDR: 0.0.0.0
CHADDR: 80:ee:73:40:6c:ad:00:00:00:00:00:00:00:00:00:00
 SNAME: .
 FNAME: .
OPTION:  53 (  1) DHCP message type         1 (DHCPDISCOVER)
OPTION:  61 (  7) Client-identifier         01:80:ee:73:40:6c:ad
OPTION:  55 (  7) Parameter Request List      1 (Subnet mask)
                                              3 (Routers)
                                             12 (Host name)
                                             15 (Domainname)
                                              6 (DNS server)
                                             33 (Static route)
                                            121 (Classless Static Route)
............................................
OPTION:  57 (  2) Maximum DHCP message size 576
```
which, IHMO, is way more appropriate, especially when the DHCP client talks to an ISP's DHCP server.

**Comment by yuwata** [comment]
https://github.com/systemd/systemd/blob/01dab40ba54f76c35b53fbfe78c9d7b340a2367f/src/libsystemd-network/sd-dhcp-client.c#L119-L120

I am not sure, but I guess when `Anonymize=` is set, `networkd` try to emulate the request sent by Windows 10.

**Comment by MelkorLord** [comment]
Ok fine then but then update the docs to make it clear for everyone. Nowadays, `anonymize` implies a notion of security/privacy and this is clearly not the case here.

The man page should mention that `Anonymize=true` emulates the request of Windows 10 with **lots** of additional parameters no user is aware of. As you point it, this information is buried in the source code, which no user, even sysadmin, is going to look at!

And also make it clear that anonymize have **nothing** to do with security in this context! Requesting **NETBIOS** and **WPAD** information is fin within local area networks but certainly **not** desirable within public networks and ISPs. At least in Europe, we're not forced to use ISP (filtering) proxies to access internet for the most part.

**Comment by yuwata** [comment]
> Ok fine then but then update the docs to make it clear for everyone. Nowadays, `anonymize` implies a notion of security/privacy and this is clearly not the case here.
> 
> The man page should mention that `Anonymize=true` emulates the request of Windows 10 with **lots** of additional parameters no user is aware of. As you point it, this information is buried in the source code, which no user, even sysadmin, is going to look at!
> 
> And also make it clear that anonymize have **nothing** to do with security in this context! Requesting **NETBIOS** and **WPAD** information is fine within local area networks but certainly **not** desirable within public networks and ISPs. At least in Europe, we're not forced to use ISP (filtering) proxies to access internet for the most part.

Updating the corresponding man page is OK. But, I cannot understand why you think `networkd` should not request such unnecessary information? Requesting redundant information does not mean exposing any private information, no?

**Comment by MelkorLord** [comment]
I didn't say it exposed private information, I said it gives a false sense of security. Like I said in my second comment : 

> Nowadays, `anonymize` implies a notion of security/privacy and this is clearly not the case here.

When `networkd` requests **WPAD** it could be used to auto-configure the browser silently, just like silently inserting `http_proxy` environment variable for exemple.

What do you think happens when a user opens up his VPN connection and realizes that all his traffic is sent trough the ISP's/Starbucks/InternetShop/whatever internal proxy because of that setting?

That's why I think the docs should emphasize on the fact that within `networkd`, the `Anonymize=true` option is simply hiding the hostname and trying to mimic Windows 10. It should also clearly state that this option have nothing to do with security per se.

I think that `Anonymize=true` was not properly thought trough and poses more threats than the false sense of security it seems to imply. This setting is perfectly OK within a well behaved LAN but certainly shady in the wild public internet. That's my opinion at least.

**Comment by yuwata** [comment]
Still I do not understand what is risky. We sending requests including NETBIOS or for WPAD, but provided information is simply dropped. Even so, still requesting them is insecure?

**Comment by MelkorLord** [comment]
Sorry for the delayed reply :-(

Ok now I see where the confusion was! If I understand correctly, the requested parameters are simply dropped and not made available to anyone like some third party app that would ask for such information via some kind of API?

This is very good then and again, should be mentioned in the docs also. In my paranoid world, it some app requests something, I expect it to use it somehow. That was the root of my concern, hence the message I posted.

So you're right, there's really no concern here but I guess this isn't clear from the documentation. I monitored/snooped the DHCP session precisely to see what `networkd` was doing. Seeing all those requested parameters which I did not explicitly asked for raised my eyebrows. I knew they were of no use since I tested in a minimal VM but I thought it would be another story in a desktop environment where everything tends to auto-configure itself and, most of the time, transparently (read: without his/her knowledge) for the user's benefit (**cough, cough**).

Finally, I then suggest to make some adjustments to the documentation to mention what `Anonymize=true` really does and to what extent. Mimicking Windows10, asking a bunch of LAN-related parameters but dropping them in the end. That should make everybody happy.

**Comment by poettering** [comment]
OK, let's convert this into a documentation bug hence.

**Comment by poettering** [comment]
(PR for this very welcome for this btw)

**Comment by poettering** [comment]
Fix pending in #11989
